src/hint_processor/hint_utils.zig

```diff
@@ -0,0 +1,114 @@
+
+pub const IdsManager = struct {
+    const Self = @This();
+
+    References: std.AutoHashMap([]const u8, HintReference),
+    HintApTracking: ApTracking,
+    AccessibleScopes: []const []const u8,
+
+    pub fn getFelt(self: *Self, name: []const u8, vm: *CairoVM) !Felt252 {
+        const val = try self.get(name, vm);
+        const felt = try vm.getFelt(val);
+        return felt;
+    }
+
+    pub fn get(self: *Self, name: []const u8, vm: *CairoVM) !?*MaybeRelocatable {
+        if (self.References.get(name)) |reference| {
+            if (getValueFromReference(reference, self.HintApTracking, vm)) |val| {
+                return val;
+            }
+        }
+        return error.ErrUnknownIdentifier;
+    }
+};
+
+pub fn getValueFromReference(reference: *HintReference, ap_tracking: ApTracking, vm: *CairoVM) ?*MaybeRelocatable {
+    // Handle the case of immediate
+    if (@typeInfo(reference.offset1) == .immediate) {
+        return MaybeRelocatable.fromFelt(Felt252.fromInt(reference.offset1.immediate));
+    }
+    const addr = getAddressFromReference(reference, ap_tracking, vm);
+    if (addr) |address| {
+        if (reference.dereference) {
+            const val = vm.segments.memory.get(address);
+            if (val) |value| {
+                return value;
+            }
+        } else {
+            return MaybeRelocatable.fromRelocatable(address);
+        }
+    }
+    return null;
+}
+
+pub fn getAddressFromReference(reference: *HintReference, ap_tracking: ApTracking, vm: *CairoVM) !?Relocatable {
+    if (@typeInfo(reference.offset1) != .reference) {
+        return null;
+    }
+    const offset1 = getOffsetValueReference(reference.offset1, reference.ap_tracking_data, ap_tracking, vm) catch return null;
+    if (offset1) |*offset1_rel| {
+        switch (reference.offset2) {
+            .reference => {
+                const offset2 = getOffsetValueReference(reference.offset2, reference.ap_tracking_data, ap_tracking, vm) catch return null;
+                if (offset2) |*offset2_val| {
+                    const res = offset1_rel.AddMaybeRelocatable(offset2_val) catch return null;
+                    return res;
+                }
+            },
+            .value => {
+                const res = offset1_rel.AddInt(reference.offset2.Value) catch return null;
+                return res;
+            },
+            else => {},
+        }
+    }
+    return null;
+}
+
+pub fn getOffsetValueReference(offset_value: OffsetValue, ref_ap_tracking: ApTracking, hint_ap_tracking: ApTracking, vm: *CairoVM) ?*MaybeRelocatable {
+    var base_addr: Relocatable = undefined;
+    var ok: bool = true;
+    switch (offset_value.reference.Register) {
+        .FP => base_addr = vm.run_context.fp,
+        .AP => {
+            const res = applyApTrackingCorrection(vm.run_context.ap, ref_ap_tracking, hint_ap_tracking);
+            if (res) |addr| {
+                base_addr = addr;
+            } else {
+                ok = false;
```

Comment:
Here `base_addr` should have a value or null wdt? so that `base_addr` is of type `?Relocatable`. Otherwise, we let it as undefined here

Reply:
I am not sure why do we need it to be optional?

Reply:
Because in this case `base_addr` still is undefined here no? There is no allocation